File: src/lib/capabilities.ts
import { harnessAdapters } from "../adapters/index.js";
import type { HarnessName } from "../domain/harness.js";

export function fieldSupport(
  kind: "skill" | "agent",
): Map<string, HarnessName[]> {
  const result = new Map<string, HarnessName[]>();
  for (const [name, adapter] of Object.entries(harnessAdapters) as Array<
    [HarnessName, (typeof harnessAdapters)[HarnessName]]
  >) {
    const keys =
      kind === "skill"
        ? adapter.capabilities.skillFrontmatterKeys
        : adapter.capabilities.agentFrontmatterKeys;
    for (const key of keys) {
      const existing = result.get(key) ?? [];
      existing.push(name);
      result.set(key, existing);
    }
  }
  return result;
}

export function eventSupport(): Map<string, HarnessName[]> {
  const result = new Map<string, HarnessName[]>();
  for (const [name, adapter] of Object.entries(harnessAdapters) as Array<
    [HarnessName, (typeof harnessAdapters)[HarnessName]]
  >) {
    for (const event of adapter.capabilities.hookEvents) {
      const existing = result.get(event) ?? [];
      existing.push(name);
      result.set(event, existing);
    }
  }
  return result;
}

export function toolSupport(): Map<string, HarnessName[]> {
  const result = new Map<string, HarnessName[]>();
  for (const [name, adapter] of Object.entries(harnessAdapters) as Array<
    [HarnessName, (typeof harnessAdapters)[HarnessName]]
  >) {
    for (const tool of adapter.capabilities.toolNames) {
      const existing = result.get(tool) ?? [];
      existing.push(name);
      result.set(tool, existing);
    }
  }
  return result;
}

File: python/cheese_flow/lib/capabilities.py
"""Port of `src/lib/capabilities.ts` — adapter capability inversion helpers."""

from __future__ import annotations

from typing import Literal

from cheese_flow.adapters import HARNESS_ADAPTERS
from cheese_flow.lib.harness import HarnessName

CapabilityKind = Literal["skill", "agent"]


def field_support(kind: CapabilityKind) -> dict[str, list[HarnessName]]:
    result: dict[str, list[HarnessName]] = {}
    for name, adapter in HARNESS_ADAPTERS.items():
        keys = (
            adapter.capabilities.skillFrontmatterKeys
            if kind == "skill"
            else adapter.capabilities.agentFrontmatterKeys
        )
        for key in keys:
            result.setdefault(key, []).append(name)
    return result


def event_support() -> dict[str, list[HarnessName]]:
    result: dict[str, list[HarnessName]] = {}
    for name, adapter in HARNESS_ADAPTERS.items():
        for event in adapter.capabilities.hookEvents:
            result.setdefault(event, []).append(name)
    return result


def tool_support() -> dict[str, list[HarnessName]]:
    result: dict[str, list[HarnessName]] = {}
    for name, adapter in HARNESS_ADAPTERS.items():
        for tool in adapter.capabilities.toolNames:
            result.setdefault(tool, []).append(name)
    return result
